src/compute/src/render/reduce.rs

```diff
@@ -961,66 +970,39 @@ where
                 }
                 let values = row_builder.clone();
 
-                row_packer = row_builder.packer();
-                row_packer.push(Datum::UInt64(values.hashed()));
-                row_packer.extend(&key);
-                (row_builder.clone(), values)
+                // Apply the initial mod here.
+                let hash = values.hashed() % first_mod;
+                let hash_key =
+                    row_builder.pack_using(std::iter::once(Datum::from(hash)).chain(&key));
+                (hash_key, values)
             });
 
             // Repeatedly apply hierarchical reduction with a progressively coarser key.
-            let pairer = Pairer::new(1);
-            for b in buckets.into_iter() {
-                let input = stage.map(move |(hash_key, values)| {
-                    let mut hash_key_iter = hash_key.iter();
-                    let hash = hash_key_iter.next().unwrap().unwrap_uint64() % b;
-                    let hash_key = pairer.merge(
-                        std::iter::once(Datum::from(hash)),
-                        hash_key_iter.take(key_arity),
-                    );
-                    (hash_key, values)
-                });
+            for (index, b) in buckets.into_iter().enumerate() {
+                // Apply subsequent bucket mods for all but the first round.
+                let input = if index == 0 {
+                    stage
+                } else {
+                    stage.map(move |(hash_key, values)| {
+                        let mut hash_key_iter = hash_key.iter();
+                        let hash = hash_key_iter.next().unwrap().unwrap_uint64() % b;
+                        let hash_key = SharedRow::pack(
+                            std::iter::once(Datum::from(hash)).chain(hash_key_iter.take(key_arity)),
```

Comment:
The `hash_key_iter.take()` seems likely to iterate through `Datum`s, and unpack and repack them. It seems like we should be able to get around this, as .. for the moment at least the state of `hash_key_iter` is probably a `&[u8]` that we could memcpy?

Reply:
Left a comment.